File: core/game-hooks/session_dialogue.c
/* @layer core-game-hooks @kind native */
// Session dialogue supplement — swaps the live dialogue blob for a host-composed one.
//
// The baked asset blob is shared per ROM across every profile, so per-seed receipt
// text can never live there. Instead the host composes a full dialogue blob at
// randomizer session start — the active language's baked lines byte-identical, plus
// one pre-rendered contextual line per planned grant — writes it to the virtual FS,
// and this file adopts it: g_zenv.dialogue_blk is repointed at a private copy, and
// restored to the baked blob when the session ends. The blob format is exactly what
// ZeldaSetLanguage selects (index 0 = dictionary, index 1 = per-line dialogue), so
// the vendored text engine reads it with no further changes.
//
// The host re-composes and re-adopts the blob mid-session whenever a line's numbers
// move (a found/total count the tracker advanced). That swap is safe at any frame
// boundary: the text engine copies a message out of the blob into its own buffer when
// the message opens (messaging.c Text_LoadCharacterBuffer) and never reads the blob
// while the box shows, so freeing the previous copy here cannot pull text from under
// a live box. The host keeps the pool's size and order, so every armed id stays valid.
//
// Ungated by design: the supplement carries the baked lines unchanged, so adopting
// it is behavior-neutral outside the randomizer's own message ids — and the message
// SUBSTITUTION stays gated at its application site (receipt_messages.c).
#include <stdlib.h>
#include "game_hooks_internal.h"
#include "src/util.h"

static const char kSessionDialoguePath[] = "session_dialogue.bin";

// The adopted copy (owned here) and the baked blob it replaced.
static uint8 *g_session_dialogue = NULL;
static MemBlk g_baked_dialogue = { NULL, 0 };

// Line count of a packed dialogue blob, for the adoption log line.
static int DialogueLineCount(MemBlk blob) {
  MemBlk lines = FindIndexInMemblk(blob, 1);
  if (lines.ptr == NULL || lines.size < 2) return 0;
  size_t mx = *(uint16 *)(lines.ptr + lines.size - 2);
  if (mx >= 8192) mx -= 8192;
  return (int)mx + 1;
}

// Adopt the host-composed blob written to the session file. Returns the blob's line
// count, or 0 when the file is missing or malformed (the baked blob stays active).
EMSCRIPTEN_KEEPALIVE
int WasmLoadSessionDialogue(void) {
  size_t length = 0;
  uint8 *data = ReadWholeFile(kSessionDialoguePath, &length);
  if (data == NULL) {
    printf("[Randomizer] Session dialogue: no %s to load\n", kSessionDialoguePath);
    return 0;
  }
  MemBlk blob = { data, length };
  if (FindIndexInMemblk(blob, 1).ptr == NULL) {
    printf("[Randomizer] Session dialogue: malformed blob (%d bytes) — keeping the baked dialogue\n",
           (int)length);
    free(data);
    return 0;
  }
  // Remember the baked blob once, before the first swap, so clear can restore it.
  if (g_baked_dialogue.ptr == NULL) g_baked_dialogue = g_zenv.dialogue_blk;
  free(g_session_dialogue);
  g_session_dialogue = data;
  g_zenv.dialogue_blk = blob;
  int lines = DialogueLineCount(blob);
  printf("[Randomizer] Session dialogue adopted: %d lines (%d bytes)\n", lines, (int)length);
  return lines;
}

// Restore the baked dialogue blob and drop the session copy.
EMSCRIPTEN_KEEPALIVE
void WasmClearSessionDialogue(void) {
  if (g_session_dialogue == NULL) return;
  if (g_baked_dialogue.ptr != NULL) g_zenv.dialogue_blk = g_baked_dialogue;
  free(g_session_dialogue);
  g_session_dialogue = NULL;
  printf("[Randomizer] Session dialogue cleared — baked dialogue restored\n");
}

File: apps/web/src/lib/game/session-dialogue/session-dialogue.ts
/* @layer bridge-wasm @kind logic */
/**
 * Session dialogue: owns the per-session pool of pre-rendered contextual
 * receipt lines and pushes them into the core's live dialogue memory. The
 * composed blob is the active language's baked bytes verbatim (dictionary and
 * every baked line, class templates included) plus one compressed line per
 * session message, written to the virtual FS and adopted by
 * WasmLoadSessionDialogue (session_dialogue.c). Per-seed text therefore never
 * touches the shared asset blob; the baked class lines at 397-401 remain the
 * fallback for any grant without a pre-rendered line.
 *
 * A line arrives as candidates, fullest first (receipt-line.type.ts); the
 * composer keeps the first one that fits the three visible rows against the
 * language's real glyph widths. The pool is re-composed whenever the session
 * re-renders its lines (the tracker moved a found/total number): the ids
 * stay put because the count and order do, and the swap is safe between
 * frames because the text engine copies a line out of the blob when a
 * message opens, never while it shows. Encoded lines are cached by text, so
 * a refresh re-encodes only what changed.
 */

import { compressStrings } from '@shared/asset-extraction/text/dialogue-encoder';
import { kLanguages } from '@shared/asset-extraction/text/data/language-data';
import {
  RANDOMIZER_MSG_BASE, randomizerTemplateTexts,
} from '@shared/asset-extraction/text/data/randomizer-templates';
import { packPackedBytes } from '@shared/asset-extraction/packed-bytes';
import { receiptLineCandidates } from '@shared/randomizer/receipt-text/receipt-line.type';
import { log } from '../../log-bus';
import { getModule } from '../wasm-bridge';
import { readActiveDialogue } from './active-dialogue';
import { fitReceiptLine, sanitizeForAlphabet, wrapMessageText } from './wrap-message';
import type { ReceiptLine } from '@shared/randomizer/receipt-text/receipt-line.type';
import type { ActiveDialogue } from './active-dialogue';

const SESSION_FILE = '/session_dialogue.bin';

let sessionLines: ReceiptLine[] = [];
let active: ActiveDialogue | null = null;
/** Message id of the first session line in the composed blob. */
let baseId: number | null = null;
/** Prepared (wrapped) text → its compressed chunk, for cheap re-composition. */
const encoded = new Map<string, Uint8Array>();

/** Baked chunks with every template line guaranteed present from 397 up. */
const baseChunksOf = (dialogue: ActiveDialogue): Uint8Array[] | null => {
  const { code, lineChunks } = dialogue;
  if (lineChunks.length < RANDOMIZER_MSG_BASE) return null;
  const templates = randomizerTemplateTexts(code);
  const baked = lineChunks.length - RANDOMIZER_MSG_BASE;
  if (baked >= templates.length) return lineChunks;
  // A blob from an older bake stops short of the newer templates: append the ones it
  // lacks so the fixed template ids stay correct below the session lines. Without this
  // a session line would sit at an id the core shows as a template line.
  return [...lineChunks, ...compressStrings([...templates.slice(baked)], code)];
};

/** The candidate that fits the box, sanitized and wrapped into line commands. */
const prepareLine = (line: ReceiptLine, alphabet: readonly string[], dialogue: ActiveDialogue): string => {
  const candidates = receiptLineCandidates(line).map((text) => sanitizeForAlphabet(text, alphabet));
  return wrapMessageText(fitReceiptLine(candidates, alphabet, dialogue.fontWidths), alphabet, dialogue.fontWidths);
};

const encodeLines = (prepared: readonly string[], code: string): Uint8Array[] => {
  const missing = [...new Set(prepared.filter((text) => !encoded.has(text)))];
  compressStrings(missing, code).forEach((chunk, i) => encoded.set(missing[i], chunk));
  return prepared.map((text) => encoded.get(text) as Uint8Array);
};

/** Compose the current pool into a blob and hand it to the core. False = kept baked. */
const compose = (): boolean => {
  const mod = getModule();
  if (!mod) return false;
  active ??= readActiveDialogue(mod);
  const dialogue = active;
  if (dialogue === null) {
    log.randomizer('[Randomizer] Session dialogue skipped: asset blob unreadable', 'warn');
    return false;
  }
  const config = kLanguages[dialogue.code];
  if (!config) {
    // A Language Studio set compiles under its own id; its base encoder is not
    // recoverable here, so the class template lines stay the fallback.
    log.randomizer(`[Randomizer] Session dialogue skipped: no encoder for language "${dialogue.code}"`, 'warn');
    return false;
  }
  const baseChunks = baseChunksOf(dialogue);
  if (baseChunks === null) {
    log.randomizer('[Randomizer] Session dialogue skipped: baked dialogue is incomplete', 'warn');
    return false;
  }
  try {
    const prepared = sessionLines.map((line) => prepareLine(line, config.alphabet, dialogue));
    const blob = packPackedBytes([
      dialogue.dictPacked,
      packPackedBytes([...baseChunks, ...encodeLines(prepared, dialogue.code)]),
    ]);
    mod.FS.writeFile(SESSION_FILE, blob);
    // Returns the adopted blob's line count (0 = refused, baked blob kept).
    const adopted = mod.ccall('WasmLoadSessionDialogue', 'number', [], []);
    if (adopted < baseChunks.length + sessionLines.length) return false;
    baseId = adopted - sessionLines.length;
    return true;
  } catch (error) {
    log.randomizer(`[Randomizer] Session dialogue compose failed: ${String(error)}`, 'warn');
    return false;
  }
};

/** Replace the pool with |lines|; returns their message ids, or null when kept baked. */
const setSessionReceiptMessages = (lines: readonly ReceiptLine[]): number[] | null => {
  sessionLines = [...lines];
  if (!compose()) return null;
  return sessionLines.map((_, i) => (baseId as number) + i);
};

/** Append one line (a runtime arrival, e.g. an online receipt); null when kept baked. */
const appendSessionReceiptMessage = (line: ReceiptLine): number | null => {
  sessionLines.push(line);
  if (!compose()) { sessionLines.pop(); return null; }
  return (baseId as number) + sessionLines.length - 1;
};

/** Drop the pool and restore the baked dialogue blob (session stop). */
const clearSessionDialogue = (): void => {
  sessionLines = [];
  active = null;
  baseId = null;
  encoded.clear();
  const mod = getModule();
  if (mod) mod.ccall('WasmClearSessionDialogue', null, [], []);
};

export { appendSessionReceiptMessage, clearSessionDialogue, setSessionReceiptMessages };
